File: cs259/project/matlab/lowpassFIR.c
// LOWPASS FIR FILTER
// lowpass_output: lowpass filtered output
// input_data: input samples to be filtered
// input_size: number of input samples
// filter_coeffs: FIR filter coefficients
// n_coeffs: number of filter coefficients

void lowpassFIR(float *lowpass_output, float *input_data, int input_size, float *filter_coeffs, int n_coeffs) {
    // different equation representation of lowpass FIR filter
    int i, k, k_limit;
    for (i=0; i<input_size; i++) {
        k_limit = (k < n_coeffs) ? k : n_coeffs;
        for (k=0; k<k_limit; k++) {
            lowpass_output[i] += filter_coeffs[k]*input_data[i-k];
        }
    }

    return;
}

File: digitalFiltersDesigns/lowpassFIR.m
function [ lowpass_output ] = lowpassFIR( input_data, filter_coeffs )
%lowpassFIR -- Direct Form Lowpass FIR filter with given coefficients
%   Produces lowpass filtered version of the input based on filte
%   coefficients passed as an argument to the function.
N_samples = length(input_data); % no. of samples of input data
N_coeffs  = length(filter_coeffs); % no. of filter coefficients

lowpass_output = zeros(1, N_samples);

% nested loop for filter difference equation
for i=1:N_samples
    for k=1:min(i,N_coeffs)
        lowpass_output(i) = lowpass_output(i) + ...
            input_data(i-k+1)*filter_coeffs(k);
    end
end

end

